indexing-service/src/main/java/org/apache/druid/indexing/worker/shuffle/DeepStorageIntermediaryDataManager.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.worker.shuffle;
+
+import com.google.common.io.ByteSource;
+import com.google.inject.Inject;
+import org.apache.druid.indexing.common.TaskToolbox;
+import org.apache.druid.indexing.common.task.batch.parallel.DeepStoragePartitionStat;
+import org.apache.druid.indexing.common.task.batch.parallel.GenericPartitionStat;
+import org.apache.druid.segment.loading.DataSegmentPusher;
+import org.apache.druid.timeline.DataSegment;
+import org.apache.druid.timeline.partition.BucketNumberedShardSpec;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+
+public class DeepStorageIntermediaryDataManager implements IntermediaryDataManager
+{
+  public static final String SHUFFLE_DATA_DIR_PREFIX = "shuffle-data";
+  private final DataSegmentPusher dataSegmentPusher;
+
+  @Override
+  public DataSegment addSegment(String supervisorTaskId, String subTaskId, DataSegment segment, File segmentDir)
+      throws IOException
+  {
+    final BucketNumberedShardSpec<?> bucketNumberedShardSpec = (BucketNumberedShardSpec<?>) segment.getShardSpec();
+    final String partitionFilePath = getPartitionFilePath(
+        supervisorTaskId,
+        subTaskId,
+        segment.getInterval(),
+        bucketNumberedShardSpec.getBucketId() // we must use the bucket ID instead of partition ID
+    );
+    return dataSegmentPusher.pushToPath(segmentDir, segment, SHUFFLE_DATA_DIR_PREFIX + "/" + partitionFilePath);
```

Comment:
The temporary zip file created will no longer be at `taskConfig.getTaskTempDir(subTaskId);` 
Will this be a problem? Should we document this change?

Reply:
I think the path `taskConfig.getTaskTempDir(subTaskId);` is more relevant for local storage and not for deep store. For deep store it would sense to have a fixed separate dir to store shuffle data so that either a coordinator duty can clean it up or the dir can be marked for auto-cleanup. Either I can document this and if required can have this prefix configurable.

Reply:
The `taskConfig.getTaskTempDir(subTaskId);` is not the local for storing the segment in local storage. The config is for the temp directory to store the zipped file, before moving the zipped file to `StorageLocation#path/supervisorTaskId/startTimeOfSegment/endTimeOfSegment/bucketIdOfSegment.`.

Deep storage also does create a zip file (same as the local storage) before copying it over to the final deep storage location. However, this temporary zip file will not be created at `taskConfig.getTaskTempDir(subTaskId);` but will be at `File.createTempFile("druid", "index.zip");`. 

I am not sure the exact purpose for `taskConfig.getTaskTempDir(subTaskId);` but this will no longer holds for the temporary zip file location before pushing to deep storage

Reply:
I see what you are saying, the temporary zip file created by `DeepStorageIntermediaryDataManager` will be cleaned up after the push so I don't think it matters.

Reply:
That was my thought too however, im not sure why is it configurable and exposed as a taskConfig. I tried looking at the PR that added this taskConfig but isn't sure. Maybe @jihoonson can confirm? Thanks!

Reply:
Hmm where will that temp file be created? My concerns are 2 folds. In general, I prefer having all temp files in one place. It will simplify the problem of cleaning them up.

1) How will the temp file be deleted if the task fails before it deletes the file? `taskTempDir` is cleaned up automatically after a task failure.
2) Users could want to use a particular disk space for ingestion temp files and allocate a reasonable amount of disk for `taskDir`.

Reply:
My thought here is simple, tasks push segments to deep storage and they use some temporary space for that. I am just using the same mechanism so it will be same lifecycle as of segment being pushed. Am I missing something here ?